SO-3125: Resolve validation performance issues

zoliMolnar opened the pull request:
https://snowowl.atlassian.net/browse/SO-3125

This pull request resolves validation performances by utilizing the jobs API, to async run multiple validation rules.

Commit:
SO-3125: Add missing package declarations to rules

Commit:
SO-3125: Add checkType property to validation-rules-common rules

Commit:
SO-3125: Add checkType enum and new property to ValidationRule

Commit:
SO-3125: Added custom ValidationJob and ValidationRuleSchedulingRule...

... implementations to make use of the eclipse job API

Commit:
SO-3125: Refactor ValidationThreadPool to use the Job API instead of ...

... the executor service

Commit:
SO-3125: Added Promise.all to ValidateRequest to wait for every job's...

... "end" in ValidateRequest changed isConflicting logic in
ValidationRuleSchedulingRule, added tests to test ValidationThreadPool

Commit:
SO-3125: Modified logic of isConflicting in ValidationRuleSchedulingRule

Commit:
SO-3125: Added rejection to promise if the job status after finishing...

... returned with not OK status

Commit:
SO-3125: Added test cases for most use-case functionalities

Commit:
SO-3125: Reduce the limit to 3 from 4 of medium checkType validation...

... rule validation

Commit:
SO-3125: Removed extra lines

cmark commented on core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationThreadPoolTest.java:
```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2011-2018 B2i Healthcare Pte Ltd, http://b2i.sg
```
New files should indicate current year as starting year in the license header.

cmark commented on core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationThreadPoolTest.java:
```diff
@@ -0,0 +1,170 @@
+ */
+public class ValidationThreadPoolTest {
+
+	private static final int MAXIMUM_AMOUNT_OF_RUNNING_EXPENSIVE_JOBS = 1;
```
Provide this as configuration to the `ValidationThreadPool` class so deployments with more computational power (CPU, RAM, ES cluster nodes) can increase the number of expensive jobs. This would be useful for the other `CheckType`s as well.

cmark commented on core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationThreadPoolTest.java:
```diff
@@ -0,0 +1,170 @@
+
+	@Test
+	public void testConcurrentExpensiveJobs() {
+		final ValidationThreadPool pool = new ValidationThreadPool(VALIDATION_THREAD_COUNT);
```
Test fixture initialization should go into a `@Before` method, please convert common variables to fields and initialize them in that method.

cmark commented on core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationThreadPoolTest.java:
```diff
@@ -0,0 +1,170 @@
+
+		final List<Promise<Object>> validationPromises = Lists.newArrayList();
+		for (int i = 0; i < 10; i++) {
+			validationPromises.add(pool.submit(CheckType.NORMAL, createValidatableRunnable(CheckType.FAST, manager)));
```
⚠️ NORMAL vs FAST!

cmark commented on core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationThreadPoolTest.java:
```diff
@@ -0,0 +1,170 @@
+				if (CheckType.NORMAL == checkType) {
+					assertTrue(runningNormalJobs <= MAXIMUM_AMOUNT_OF_RUNNING_NORMAL_JOBS);
+				}
+				Thread.sleep(500);
```
Would be nice to configure this hardcoded 500 ms timeout value based on the given `CheckType`, just to simulate the cost more accurately.

cmark commented on core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationThreadPoolTest.java:
```diff
@@ -0,0 +1,170 @@
+				}
+				Thread.sleep(500);
+			} catch (InterruptedException e) {
+				e.printStackTrace();
```
Wrap all non-`RuntimeException`s into `RuntimeException`s and let them broke the tests.

Commit:
Merge remote-tracking branch 'origin/6.x' into
issue/SO-3125_resolve_validation_performance_issues

Conflicts:
	snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/ruleSnomedCommon1.groovy
	snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/ruleSnomedCommon2.groovy

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationJob.java:
```diff
@@ -0,0 +1,54 @@
+	private final Object family;
+	private final Runnable runnable;
+	
+	public ValidationJob(Object family, Runnable runnable) {
```
Add a null check to `family` and `runnable` arguments

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationJob.java:
```diff
@@ -0,0 +1,54 @@
+	
+	@Override
+	protected IStatus run(IProgressMonitor monitor) {
+		runnable.run();
```
Add catch clause for `OperationCancelledException` and return CANCEL_STATUS.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationRuleSchedulingRule.java:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2011-2018 B2i Healthcare Pte Ltd, http://b2i.sg
```
⚠️ Fix license year!

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationJob.java:
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2011-2018 B2i Healthcare Pte Ltd, http://b2i.sg
```
⚠️ Fix license year!

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationJob.java:
```diff
@@ -0,0 +1,54 @@
+	
+	@Override
+	public boolean belongsTo(Object family) {
+		return (this.family == null || family == null) ? false : this.family.equals(family);
```
Use `java.util.Objects.equals()`.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationRuleSchedulingRule.java:
```diff
@@ -0,0 +1,107 @@
+	
+	private final CheckType checkType;
+	private final int maxConcurrentJobs;
+	private final String uniqueRuleId;
```
What is `uniqueRuleId`? This should be either just `ruleId` or just `id`. IDs are meant to be **unique**.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationRuleSchedulingRule.java:
```diff
@@ -0,0 +1,107 @@
+	
+	@Override
+	public int hashCode() {
+		int hash = 17;
```
Use `Objects.hash(args)` correctly, where you pass all values directly to the function and not one-by-one.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationRuleSchedulingRule.java:
```diff
@@ -0,0 +1,107 @@
+	
+	@Override
+	public String toString() {
+		return uniqueRuleId;
```
Add `checkType` to `toString()`

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationThreadPool.java:
```diff
@@ -15,42 +15,50 @@
+		job.setRule(schedulingRule);
+		job.addJobChangeListener(new JobChangeAdapter() {
+
+			public void done(IJobChangeEvent event) {
```
Missing `@Override` annotation.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationThreadPool.java:
```diff
@@ -15,42 +15,50 @@
+		final Job job = new ValidationJob(checkType.getName(), runnable);
+		String uniqueRuleId = UUID.randomUUID().toString();
+		final ISchedulingRule schedulingRule = new ValidationRuleSchedulingRule(checkType, maxValidationThreadCount, uniqueRuleId);
+		final Promise<Object> isRuleDonePromise = new Promise<>();
```
Rename `isRuleDonePromise` to just `promise`.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/validation/rule/ValidationRuleCreateRequest.java:
```diff
@@ -44,6 +41,8 @@
 	@NotNull 
 	private Severity severity;
 	
+	private CheckType checkType;
```
Add `NotNull` annotation or default value `CheckType.NORMAL`.

Commit:
SO-3125: Add catch clause for OperationCanceledException to return...

... cancel status
- Add check against null values for family and runnables, change
belongsTo to use Objects.equal

- Add NotNull annotaion to ValidationRuleCreateRequest

- Add configurability to limit the maximum amount of concurrent normal
or expensive jobs to ValidationBootstrap, ValidationThreadPool,
ValidationRuleSchedulingRule

- Fix test class and added logging for test

Commit:
SO-3125: Fix test classes

Commit:
SO-3125: Refactor aggregation in ruleSnomedCommon2 to use fields

Commit:
SO-3125: Increase scroll limit to 50_000, prevent fetching of already...

... fetched concepts in SnomedValidationIssueDetailExtension.class,
added validation rule execution time logging to ValidateRequest.class

Commit:
SO-3125: Remove unused imports

Commit:
Merge remote-tracking branch 'origin/6.x' into issue/SO-3125_resolve_validation_performance_issues

cmark committed:
Merge remote-tracking branch 'origin/6.x' into issue/SO-3125_resolve_validation_performance_issues

cmark committed:
SO-3125: reduce memory consumption of SnomedQueryValidationRuleEvaluator

cmark committed:
SO-3125: use a queue to persist issues generated so far by workers

cmark committed:
SO-3125: configure top hits aggregation to skip stored field loading...

...in case of _source = false configuration

This will reduce the memory consumption of aggregations with top hits
somewhat.

cmark committed:
SO-3125: improve validate request log messages

cmark commented on core/com.b2international.snowowl.datastore.server.tests/src/com/b2international/snowowl/core/validation/ValidationThreadPoolTest.java:
```diff
@@ -54,13 +67,12 @@ public void testConcurrentExpensiveJobs() {
 		}
 
 		Promise.all(validationPromises).getSync();
+		System.out.println();
```
Remove `println` calls. If you would like to separate the log output of the test cases then use a `JUnit Rule` like this class: `com.b2international.snowowl.test.commons.TestMethodNameRule`.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationBootstrap.java:
```diff
@@ -78,7 +78,11 @@ public void preRun(SnowOwlConfiguration configuration, Environment env) throws E
 			// TODO make this configurable
 			int numberOfValidationThreads = Math.max(2, Runtime.getRuntime().availableProcessors() / 2);
-			env.services().registerService(ValidationThreadPool.class, new ValidationThreadPool(numberOfValidationThreads));
+			// TODO calculate these based on systems overall strength
```
Move these to actual `snowowl_config.yml` configuration values.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationJob.java:
```diff
@@ -48,7 +60,7 @@ protected IStatus run(IProgressMonitor monitor) {
 	@Override
 	public boolean belongsTo(Object family) {
-		return (this.family == null || family == null) ? false : this.family.equals(family);
+		return Objects.equal(this.family, family);
```
Use `java.util.Objects`, please.

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationRuleSchedulingRule.java:
```diff
@@ -83,9 +92,7 @@ public boolean isConflicting(ISchedulingRule schedulingRule) {
-		hash = hash * 31 + Objects.hash(uniqueRuleId); 
+		hash = hash * 31 + Objects.hash(checkType, id, maxConcurrentJobs, maxConcurrentExpensiveJobs);
 		
 		return hash;
```
Just return the value computed by the Objects.hash method, like this:

```
@Override
public int hashCode() {
    return Objects.hash(checkType, id, maxConcurrentJobs, maxConcurrentExpensiveJobs);
}
```

cmark commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/internal/validation/ValidationRuleSchedulingRule.java:
```diff
@@ -96,12 +103,12 @@ public boolean equals(Object obj) {
 	@Override
 	public String toString() {
-		return uniqueRuleId;
+		return id + " " + checkType.getName();
```
Use `String.format()` method.

Commit:
SO-3125: Make the amount of validation threads, concurrent normal...

... and expensive validation jobs configurable

Commit:
SO-3125: Make validation work with issue deltas instead of dropping...

... and persisting them again

Commit:
SO-3125: Use String.format in toString

Commit:
SO-3125: Remove printlns from test use ValidationTestMethodNameRule...

... to separate tests in the log

Commit:
SO-3125: Use java.util.Objects in ValidationJob return Objects.hash...

... computed value in ValidationRuleSchedulingRule hashCode

cmark committed:
SO-3125: use Stopwatch instead of custom elapsed time tracking

cmark reviewed (approved):
LGTM!

zoliMolnar commented:
Thank you!
